subsys/bluetooth/mesh/vnd/dm_srv.c

```diff
@@ -0,0 +1,457 @@
+		       struct net_buf_simple *buf)
+{
+	int err;
+	struct bt_mesh_dm_srv *srv = model->user_data;
+
+	if (srv->is_busy) {
+		return -EBUSY;
+	}
+
+	srv->target_addr = ctx->addr;
+
+	enum dm_ranging_mode method = net_buf_simple_pull_u8(buf);
+	uint8_t timeout = net_buf_simple_pull_u8(buf);
+	struct dm_request req = {
+		.role = DM_ROLE_INITIATOR,
+		.ranging_mode = method,
+		.access_address = (ctx->addr << 16) + ctx->recv_dst,
+		.start_delay_us = 0
+	};
+
+	srv->is_busy = true;
+	k_sem_reset(&srv->dm_ready_sem);
+	err = dm_request_add(&req);
+	if (err) {
+		BT_ERR("DM add request failed: %d", err);
+		srv->is_busy = false;
+		return err;
+	}
+
+	srv->current_role = DM_ROLE_INITIATOR;
+	k_work_reschedule(&srv->timeout, K_MSEC(timeout * 10));
+
+	return 0;
+}
+
+const struct bt_mesh_model_op _bt_mesh_dm_srv_op[] = {
+	{
+		BT_MESH_DM_CONFIG_OP,
+		BT_MESH_LEN_MIN(BT_MESH_DM_CONFIG_MSG_LEN_MIN),
+		handle_cfg,
+	},
+	{
+		BT_MESH_DM_START_OP,
+		BT_MESH_LEN_MIN(BT_MESH_DM_START_MSG_LEN_MIN),
+		handle_start,
+	},
+	{
+		BT_MESH_DM_RESULT_GET_OP,
+		BT_MESH_LEN_MIN(BT_MESH_DM_RESULT_GET_MSG_LEN),
+		handle_result_get,
+	},
+	{
+		BT_MESH_DM_RESULT_SYNC_OP,
+		BT_MESH_LEN_MIN(BT_MESH_DM_RESULT_SYNC_MSG_LEN),
+		handle_sync,
+	},
+	BT_MESH_MODEL_OP_END,
+};
+
+static void data_ready(struct dm_result *result)
+{
+	if (!dm_srv->is_busy) {
+		BT_WARN("Unexpected DM result ready, discarding");
+		return;
+	}
+
+	BT_INFO("New DM result ready for target node %04x", dm_srv->target_addr);
+
+	new_entry_store(result);
+	printk("VALUE: %f\n", result->dist_estimates.rtt.rtt);
+	if (k_work_delayable_is_pending(&dm_srv->timeout)) {
+		k_work_cancel_delayable(&dm_srv->timeout);
+		if (dm_srv->current_role == DM_ROLE_REFLECTOR) {
+			result_status_send(dm_srv->model, &dm_srv->rsp_ctx, 1, 0);
+		}
+	}
+
+	dm_srv->is_busy = false;
+}
+
+static struct dm_cb dm_cb = {
+	.data_ready = data_ready,
+};
+
+static void timeout_work(struct k_work *work)
+{
+	BT_ERR("DM attempt timed out");
+	if (dm_srv->current_role == DM_ROLE_REFLECTOR) {
+		result_status_send(dm_srv->model, &dm_srv->rsp_ctx, 1, ETIMEDOUT);
+	}
+
+	struct dm_result result = {
+		.quality = DM_QUALITY_DO_NOT_USE,
+		.status = false
+	};
+
+	new_entry_store(&result);
+
+	dm_srv->is_busy = false;
+}
+
+static int bt_mesh_dm_srv_init(struct bt_mesh_model *model)
+{
+	int err;
+	struct bt_mesh_dm_srv *srv = model->user_data;
+
+	srv->model = model;
+	dm_srv = srv;
+	k_sem_init(&srv->dm_ready_sem, 0, 1);
+
+	struct dm_init_param init_param = { .cb = &dm_cb};
+
+	err = dm_init(&init_param);
+	if (err) {
+		BT_ERR("Failed to initialize DM: %d", err);
+		return err;
+	}
+
+	k_work_init_delayable(&srv->timeout, timeout_work);
+
+#if CONFIG_BT_SETTINGS
+	k_work_init_delayable(&srv->store_timer, store_timeout);
+#endif
+	return 0;
+}
+
+static void bt_mesh_dm_srv_reset(struct bt_mesh_model *model)
+{
+	struct bt_mesh_dm_srv *srv = model->user_data;
+
+	srv->cfg.ttl = CONFIG_BT_MESH_DEFAULT_TTL;
```

Comment:
BT_MESH_DEFAULT_TTL has range from 0 -128. But we are excluding 1 and larger than 127.

Reply:
I'll add a own default for it

Reply:
I think you should specify '0' as a default here because that is the most common use case.

Reply:
It is 👍